Performance test for Config API

shivamkm07 opened the pull request:
# Description

This PR adds perf test for config API.
## Issue reference

<!--
We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.
-->

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

mukundansundar commented on tests/apps/perf/configuration/app.go:
```diff
@@ -0,0 +1,523 @@
+/*
+Copyright 2021 The Dapr Authors
```
nitpick ...all new files to have latest year

mukundansundar commented on tests/apps/perf/configuration/app.go:
```diff
@@ -0,0 +1,523 @@
+	if p != "" && p != "0" {
+		appPort, _ = strconv.Atoi(p)
+	}
+} // indexHandler is the handler for root path
```
nitpick live next line blank

mukundansundar commented on tests/apps/perf/configuration/app.go:
```diff
@@ -0,0 +1,523 @@
+	m := make([]interface{}, 0, 2*len(items)+1)
+	for key, item := range items {
+		val := item.Value + separator + item.Version
+		m = append(m, key, val)
```
nitpick: Not needed in test code, but avoid append and use indexing where possible.

mukundansundar commented on tests/apps/perf/configuration/app.go:
```diff
@@ -0,0 +1,523 @@
+}
+
+// extract value and version from redis value
+func getRedisValueAndVersion(redisValue string) (string, string) {
```
same as above.

shivamkm07 commented:
@mukundansundar Can you run perf test on this PR to validate the changes?

mukundansundar commented:
@shivamkm07 can you fix the conflicts?

mukundansundar commented:
/ok-to-perf

mukundansundar commented:
@shivamkm07 Please add the baseline results and the difference to the summary table if possible. And where are the logs written for the same?

shivamkm07 commented:
Latency numbers for config API:

TestConfigurationGetHTTPPerformance (229.00s)
dapr p95 latency: 81.65ms, baseline p95 latency: 19.16ms
added p95 latency: 62.49ms(326.06%)
dapr avg latency: 40.09ms, baseline avg latency: 6.91ms
added avg latency: 33.18ms(480.03%)

TestConfigurationGetGRPCPerformance (346.56s)
dapr p95 latency: 65.90ms, baseline p95 latency: 17.18ms
added p95 latency: 48.72ms(283.52%)
dapr avg latency: 31.68ms, baseline avg latency: 6.02ms
added avg latency: 25.66ms(426.28%)

TestConfigurationSubscribeHTTPPerformance (351.67s)
dapr p95 latency: 462.17ms, baseline p95 latency: 199.60ms
added p95 latency: 262.57ms(131.55%)
dapr avg latency: 236.09ms, baseline avg latency: 107.58ms
added avg latency: 128.51ms(119.46%)

TestConfigurationSubscribeGRPCPerformance (351.71s)
dapr p95 latency: 199.17ms, baseline p95 latency: 188.29ms
added p95 latency: 10.88ms(5.78%)
dapr avg latency: 106.26ms, baseline avg latency: 103.93ms
added avg latency: 2.33ms(2.24%)

mukundansundar commented on tests/perf/configuration/configuration_test.go:
```diff
@@ -0,0 +1,268 @@
+
+const (
+	numHealthChecks                   = 60 // Number of times to check for endpoint health per app.
+	defaultConfigGetThresholdMs       = 500
+	defaultConfigSubscribeThresholdMs = 1000
```
based on the latencies stated [here](https://github.com/dapr/dapr/pull/6247#issuecomment-1511028152), why do we have such high buffer values?

mukundansundar commented:
/ok-to-perf

mukundansundar commented:
@shivamkm07 the config perf tests seem to be failing for the second run

mukundansundar commented:
/ok-to-perf configuration

mukundansundar commented:
/ok-to-perf configuration

mukundansundar commented:
/ok-to-perf configuration

daixiang0 commented:
/ok-to-perf

mukundansundar commented:
/ok-to-perf configuration

mukundansundar commented:
/ok-to-perf

mukundansundar commented:
/ok-to-perf configuration

mukundansundar commented:
/ok-to-perf configuration

mukundansundar commented:
/ok-to-test configuration

mukundansundar commented:
/ok-to-perf configuration
